File: PerformanceCheckTests/CoreDataTests.m
//
//  PerfomanceCheckTests.m
//  PerformanceCheck
//
//  Created by mono on 11/30/14.
//  Copyright (c) 2014 mono. All rights reserved.
//

#import <XCTest/XCTest.h>
#import "CoreData+MagicalRecord.h"
#import <CoreData/CoreData.h>
#import "PerformanceCheckTests-Swift.h"

@interface CoreDataTests2 : XCTestCase

@end

@implementation CoreDataTests2

- (void)setUp {
    [super setUp];
    [MagicalRecord setupCoreDataStack];
    NSUInteger memberCount = 100000;
    NSManagedObjectContext* moc = [NSManagedObjectContext MR_contextForCurrentThread];
    
    for (NSUInteger i = 0; i <memberCount; i++) {
        [Member MR_createEntity];
        [moc MR_saveToPersistentStoreAndWait];
    }
}

- (void)tearDown {
    [Member MR_deleteAllMatchingPredicate:nil];
    NSManagedObjectContext* moc = [NSManagedObjectContext MR_contextForCurrentThread];
    [moc MR_saveToPersistentStoreAndWait];
    [super tearDown];
}

- (void)testCoreData {
    [self measureBlock:^{
        NSManagedObjectContext* moc = [NSManagedObjectContext MR_contextForCurrentThread];
        NSArray* members = [Member MR_findAll];
        for (Member* m in members) {
            m.loveIce = YES;
        }
        [moc MR_saveToPersistentStoreAndWait];
    }];
}

- (void)testCoreDataBatch {
    [self measureBlock:^{
        NSManagedObjectContext* moc = [NSManagedObjectContext MR_contextForCurrentThread];
        NSEntityDescription *entityDescription = [NSEntityDescription entityForName:@"Member" inManagedObjectContext:moc];
        
        NSBatchUpdateRequest *batchUpdateRequest = [[NSBatchUpdateRequest alloc] initWithEntity:entityDescription];
        
        [batchUpdateRequest setResultType:NSUpdatedObjectIDsResultType];
        [batchUpdateRequest setPropertiesToUpdate:@{ @"loveIce" : @YES }];
        
        // Execute Batch Request
        NSError *batchUpdateRequestError = nil;
        [moc executeRequest:batchUpdateRequest error:&batchUpdateRequestError];
        if (batchUpdateRequestError) {
            XCTFail(@"%@", batchUpdateRequestError);
        }
    }];
}

@end

File: PerformanceCheckTests/CoreDataTests.swift
//
//  PerformanceCheckTests.swift
//  PerformanceCheckTests
//
//  Created by mono on 11/30/14.
//  Copyright (c) 2014 mono. All rights reserved.
//

import UIKit
import XCTest

class CoreDataTests: XCTestCase {
    
    override func setUp() {
        MagicalRecord.setupCoreDataStack()
        let memberCount = 100000
        let moc = NSManagedObjectContext.MR_contextForCurrentThread()
        
        for i in 0..<memberCount {
            Member.MR_createEntity()
            moc.MR_saveToPersistentStoreAndWait()
        }
        super.setUp()
    }
    
    override func tearDown() {
        Member.MR_deleteAllMatchingPredicate(nil)
        super.tearDown()
    }
    
    func testCoreData() {
        self.measureBlock() {
            let moc = NSManagedObjectContext.MR_contextForCurrentThread()
            let members = Member.MR_findAll() as [Member]
            for m in members {
                m.loveIce = true
            }
            moc.MR_saveToPersistentStoreAndWait()
        }
    }
    
    func testCoreDataBatch() {
        self.measureBlock() {
            let moc = NSManagedObjectContext.MR_contextForCurrentThread()
            let entityDescription = NSEntityDescription.entityForName("Member", inManagedObjectContext: moc)!
            let batchUpdateRequest = NSBatchUpdateRequest(entity: entityDescription)
            batchUpdateRequest.resultType = NSBatchUpdateRequestResultType.UpdatedObjectIDsResultType
            batchUpdateRequest.propertiesToUpdate = ["loveIce": true]
            // Execute Batch Request
            var batchUpdateRequestError: NSError?
            moc.executeRequest(batchUpdateRequest, error:&batchUpdateRequestError)
            if let e = batchUpdateRequestError {
                XCTFail(e.description)
            }
        }
    }
}
